src/compute/src/render/mod.rs

```diff
@@ -1053,3 +1053,61 @@ where
         }
     })
 }
+
+/// Helper to merge pairs of datum iterators into a row or split a datum iterator
+/// into two rows, given the arity of the first component.
+#[derive(Clone, Copy, Debug)]
+struct Pairer {
+    split_arity: usize,
+}
+
+impl Pairer {
+    /// Creates a pairer with knowledge of the arity of first component in the pair.
+    fn new(split_arity: usize) -> Self {
+        Self { split_arity }
+    }
+
+    /// Merges a pair of datum iterators creating a `Row` instance.
+    fn merge<'a, I1, I2>(&self, first: I1, second: I2) -> Row
+    where
+        I1: IntoIterator<Item = Datum<'a>>,
+        I2: IntoIterator<Item = Datum<'a>>,
+    {
+        let binding = SharedRow::get();
+        let mut row_builder = binding.borrow_mut();
+        let mut row_packer = row_builder.packer();
+        row_packer.extend(first);
+        row_packer.extend(second);
+        row_builder.clone()
+    }
+
+    /// Splits a datum iterator into a pair of `Row` instances.
+    fn split<'a>(&self, datum_iter: impl IntoIterator<Item = Datum<'a>>) -> (Row, Row) {
+        let mut datum_iter = datum_iter.into_iter();
+        let binding = SharedRow::get();
+        let mut row_builder = binding.borrow_mut();
+        let mut row_packer = row_builder.packer();
+        row_packer.extend(datum_iter.by_ref().take(self.split_arity));
+        let first = row_builder.clone();
+        row_packer = row_builder.packer();
+        row_packer.extend(datum_iter);
+        (first, row_builder.clone())
+    }
+
+    /// Converts a pair of datum iterators into one of `Row` instances.
+    fn to_rows<'a, I1, I2>(&self, first: I1, second: I2) -> (Row, Row)
```

Comment:
This doesn't seem to use `self`? It seems odd to have this function, I think it's better to keep the code in-line, or move it to a different free-standing function.

Reply:
A more idiomatic approach might be to implement `FromIterator` for Row:
```rust
impl<'a> FromIterator<Datum<'a>> for Row {
    fn from_iter<T: IntoIterator<Item = Datum<'a>>>(iter: T) -> Self {
        let binding = SharedRow::get();
        let mut builder = binding.borrow_mut();
        builder.packer().extend(iter);
        builder.clone()
    }
}
```

Reply:
I really like the suggestion of implementing `FromIterator` for `Row`, but it exceeds the scope of this PR (it would have many more uses beyond in top-k and bucketed reductions). I am going to try that out on a separate PR on Friday. For now, I am just dropping the last commit and leaving the code in-line here.